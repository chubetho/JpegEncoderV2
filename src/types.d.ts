interface Block {
  R: Uint8Array
  G: Uint8Array
  B: Uint8Array
  Y: Int32Array
  Cb: Int32Array
  Cr: Int32Array
}

interface Image {
  blocks: Block[]
  metadata: {
    imageWidth: number
    imageHeight: number
    maxColor: number
    format: string
    blockWidth: number
    blockHeight: number
  }
}

interface HuffmanTable {
  offsets: Uint8Array
  symbols: Uint8Array
  codes: Uint16Array
  set?: boolean
}
