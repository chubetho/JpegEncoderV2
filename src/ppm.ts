import * as fs from 'node:fs'

export function readPpm(path: string) {
  const content = fs.readFileSync(path, 'ascii')
  const lines = content.split('\n')
  const [format, size, _maxColor, ...rest] = lines
  const [imageWidth, imageHeight] = size.trim().split(' ').map(Number)
  const maxColor = Number.parseInt(_maxColor)

  if (format !== 'P3')
    throw new Error ('Invalid ppm')

  if (maxColor !== 255)
    throw new Error ('Invalid ppm')

  const image = new Uint8Array(imageWidth * imageHeight * 3)
  let index = 0

  for (let l = 0; l < rest.length; l++) {
    const trim = rest[l].trim()
    if (trim[0] === '#')
      continue

    const row = trim.split(' ')
    for (let i = 0; i < row.length; i++) {
      if (!row[i])
        continue
      const v = Number.parseInt(row[i])
      if (Number.isNaN(v) || v < 0 || v > maxColor)
        throw new Error ('Invalid ppm')

      image[index++] = v
    }
  }

  const blockHeight = ~~((imageHeight + 7) / 8)
  const blockWidth = ~~((imageWidth + 7) / 8)
  const blocks = Array.from({ length: blockHeight * blockWidth }, () => ({
    R: new Uint8Array(64),
    G: new Uint8Array(64),
    B: new Uint8Array(64),
    Y: new Int32Array(64),
    Cb: new Int32Array(64),
    Cr: new Int32Array(64),
  }))

  const clamp = (x: number) => x < -128 ? -128 : x > 127 ? 127 : x
  for (let h = 0; h < imageHeight; h++) {
    const blockRow = ~~(h / 8)
    const pixelRow = h % 8
    for (let w = 0; w < imageWidth; w++) {
      const blockColumn = ~~(w / 8)
      const pixelColumn = w % 8
      const blockIndex = blockRow * blockWidth + blockColumn
      const pixelIndex = pixelRow * 8 + pixelColumn

      const index = (h * imageWidth + w) * 3
      const r = image[index]
      const g = image[index + 1]
      const b = image[index + 2]

      blocks[blockIndex].R[pixelIndex] = r
      blocks[blockIndex].G[pixelIndex] = g
      blocks[blockIndex].B[pixelIndex] = b

      const y = 0.299 * r + 0.587 * g + 0.114 * b - 128
      const cb = -0.1687 * r + -0.3312 * g + 0.5 * b
      const cr = 0.5 * r + -0.4186 * g + -0.0813 * b

      blocks[blockIndex].Y[pixelIndex] = clamp(y)
      blocks[blockIndex].Cb[pixelIndex] = clamp(cb)
      blocks[blockIndex].Cr[pixelIndex] = clamp(cr)
    }
  }

  return {
    blocks,
    metadata: {
      imageHeight,
      imageWidth,
      format,
      maxColor,
      blockHeight,
      blockWidth,
    },
  }
}

export function writePpm(path: string, { blocks, metadata }: Image) {
  let data = `${metadata.format}\n${metadata.imageWidth} ${metadata.imageHeight}\n${metadata.maxColor}\n`

  for (let h = 0; h < metadata.imageHeight; h++) {
    const blockRow = ~~(h / 8)
    const pixelRow = h % 8
    for (let w = 0; w < metadata.imageWidth; w++) {
      const blockColumn = ~~(w / 8)
      const pixelColumn = w % 8
      const blockIndex = blockRow * metadata.blockWidth + blockColumn
      const pixelIndex = pixelRow * 8 + pixelColumn

      const r = blocks[blockIndex].R[pixelIndex]
      const g = blocks[blockIndex].G[pixelIndex]
      const b = blocks[blockIndex].B[pixelIndex]
      data += `${r} ${g} ${b}  `
    }

    data += '\n'
  }

  fs.writeFileSync(path, data)
}
