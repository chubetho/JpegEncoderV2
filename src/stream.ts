export function useStream() {
  let nextBit = 0
  const data: number[] = []

  function writeBit(v: number) {
    if (nextBit === 0)
      data.push(v)

    data[data.length - 1] |= (v & 1) << (7 - nextBit)
    nextBit = (nextBit + 1) % 8
    if (nextBit === 0 && data[data.length - 1] === 0xFF)
      data.push(0)
  }

  function writeBits(v: number, length: number) {
    for (let i = 1; i <= length; i++)
      writeBit(v >> (length - 1))
  }

  function writeByte(v: number) {
    data.push(v)
  }

  function writeWord(v: number) {
    writeByte((v >> 8) & 0xFF)
    writeByte(v & 0xFF)
  }

  return { data, writeBit, writeBits, writeByte, writeWord }
}
