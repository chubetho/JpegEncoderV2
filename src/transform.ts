const sqrt2 = Math.SQRT2
const one_sqrt2 = 1 / sqrt2
const cos = Math.cos
const round = Math.round
const pi = Math.PI
const pi_16 = pi / 16

const m0 = cos(2 * pi_16)
const m1 = cos(4 * pi_16)
const m5 = cos(6 * pi_16)
const m2 = m0 - m5
const m3 = m1
const m4 = m0 + m5

const s0 = 0.5 * one_sqrt2
const s1 = 1 / (4 * cos(1 * pi_16))
const s2 = 1 / (4 * cos(2 * pi_16))
const s3 = 1 / (4 * cos(3 * pi_16))
const s4 = 1 / (4 * cos(4 * pi_16))
const s5 = 1 / (4 * cos(5 * pi_16))
const s6 = 1 / (4 * cos(6 * pi_16))
const s7 = 1 / (4 * cos(7 * pi_16))

export function dct(X: number[]) {
  const C = (x: number) => x === 0 ? one_sqrt2 : 1
  const Y: number[] = []
  for (let i = 0; i < 8; i++) {
    for (let j = 0; j < 8; j++) {
      let sum = 0
      for (let x = 0; x < 8; x++) {
        for (let y = 0; y < 8; y++)
          sum += X[x * 8 + y] * cos((2 * x + 1) * i * pi_16) * cos((2 * y + 1) * j * pi_16)
      }
      sum = 1 / 4 * C(i) * C(j) * sum
      Y[i * 8 + j] = sum
      sum = 0
    }
  }
  return Y
}

export function idct(X: number[]) {
  const C = (x: number) => x === 0 ? one_sqrt2 : 1
  const Y: number[] = []
  for (let x = 0; x < 8; x++) {
    for (let y = 0; y < 8; y++) {
      let sum = 0
      for (let i = 0; i < 8; i++) {
        for (let j = 0; j < 8; j++)
          sum += C(i) * C(j) * X[i * 8 + j] * cos((2 * x + 1) * i * pi_16) * cos((2 * y + 1) * j * pi_16)
      }
      Y[x * 8 + y] = round(1 / 4 * sum)
      sum = 0
    }
  }
  return Y
}

export function aan(X: number[]) {
  for (let i = 0; i < 8; i++) {
    const b0 = X[i * 8 + 0] + X[i * 8 + 7]
    const b1 = X[i * 8 + 1] + X[i * 8 + 6]
    const b2 = X[i * 8 + 2] + X[i * 8 + 5]
    const b3 = X[i * 8 + 3] + X[i * 8 + 4]
    const b4 = X[i * 8 + 3] - X[i * 8 + 4]
    const b5 = X[i * 8 + 2] - X[i * 8 + 5]
    const b6 = X[i * 8 + 1] - X[i * 8 + 6]
    const b7 = X[i * 8 + 0] - X[i * 8 + 7]

    const c0 = b0 + b3
    const c1 = b1 + b2
    const c2 = -b2 + b1
    const c3 = -b3 + b0
    const c4 = -b4 - b5
    const c5 = b5 + b6
    const c6 = b6 + b7
    const c7 = b7

    const d0 = c0 + c1
    const d1 = -c1 + c0
    const d2 = c2 + c3
    const d3 = c3
    const d4 = c4
    const d5 = c5
    const d6 = c6
    const d7 = c7
    const d8 = d4 + d6

    const e8 = d8 * m5
    const e0 = d0
    const e1 = d1
    const e2 = d2 * m1
    const e3 = d3
    const e4 = -d4 * m2 - e8
    const e5 = d5 * m3
    const e6 = d6 * m4 - e8
    const e7 = d7

    const f0 = e0
    const f1 = e1
    const f2 = e2 + e3
    const f3 = e3 - e2
    const f4 = e4
    const f5 = e5 + e7
    const f6 = e6
    const f7 = e7 - e5

    const g0 = f0
    const g1 = f1
    const g2 = f2
    const g3 = f3
    const g4 = f4 + f7
    const g5 = f5 + f6
    const g6 = f5 - f6
    const g7 = f7 - f4

    X[i * 8 + 0] = g0 * s0
    X[i * 8 + 1] = g5 * s1
    X[i * 8 + 2] = g2 * s2
    X[i * 8 + 3] = g7 * s3
    X[i * 8 + 4] = g1 * s4
    X[i * 8 + 5] = g4 * s5
    X[i * 8 + 6] = g3 * s6
    X[i * 8 + 7] = g6 * s7
  }

  for (let i = 0; i < 8; i++) {
    const b0 = X[0 * 8 + i] + X[7 * 8 + i]
    const b1 = X[1 * 8 + i] + X[6 * 8 + i]
    const b2 = X[2 * 8 + i] + X[5 * 8 + i]
    const b3 = X[3 * 8 + i] + X[4 * 8 + i]
    const b4 = X[3 * 8 + i] - X[4 * 8 + i]
    const b5 = X[2 * 8 + i] - X[5 * 8 + i]
    const b6 = X[1 * 8 + i] - X[6 * 8 + i]
    const b7 = X[0 * 8 + i] - X[7 * 8 + i]

    const c0 = b0 + b3
    const c1 = b1 + b2
    const c2 = -b2 + b1
    const c3 = -b3 + b0
    const c4 = -b4 - b5
    const c5 = b5 + b6
    const c6 = b6 + b7
    const c7 = b7

    const d0 = c0 + c1
    const d1 = -c1 + c0
    const d2 = c2 + c3
    const d3 = c3
    const d4 = c4
    const d5 = c5
    const d6 = c6
    const d7 = c7
    const d8 = d4 + d6

    const e8 = d8 * m5
    const e0 = d0
    const e1 = d1
    const e2 = d2 * m1
    const e3 = d3
    const e4 = -d4 * m2 - e8
    const e5 = d5 * m3
    const e6 = d6 * m4 - e8
    const e7 = d7

    const f0 = e0
    const f1 = e1
    const f2 = e2 + e3
    const f3 = e3 - e2
    const f4 = e4
    const f5 = e5 + e7
    const f6 = e6
    const f7 = e7 - e5

    const g0 = f0
    const g1 = f1
    const g2 = f2
    const g3 = f3
    const g4 = f4 + f7
    const g5 = f5 + f6
    const g6 = f5 - f6
    const g7 = f7 - f4

    X[0 * 8 + i] = g0 * s0
    X[1 * 8 + i] = g5 * s1
    X[2 * 8 + i] = g2 * s2
    X[3 * 8 + i] = g7 * s3
    X[4 * 8 + i] = g1 * s4
    X[5 * 8 + i] = g4 * s5
    X[6 * 8 + i] = g3 * s6
    X[7 * 8 + i] = g6 * s7
  }

  return X
}

export function sep(X: number[][]) {
  const A: number[][] = []
  for (let k = 0; k < 8; k++) {
    const c0 = k === 0 ? one_sqrt2 : 1
    A[k] = []
    for (let n = 0; n < 8; n++)
      A[k][n] = c0 * 0.5 * cos((2 * n + 1) * k * pi_16)
  }

  return dot(A, dot(X, transpose(A)))
}

function transpose(matrix: number[][]) {
  const transposed: number[][] = []
  for (let rowIdx = 0; rowIdx < 8; rowIdx++) {
    for (let colIdx = 0; colIdx < 8; colIdx++) {
      transposed[colIdx] ||= []
      transposed[colIdx][rowIdx] = matrix[rowIdx][colIdx]
    }
  }
  return transposed
}

function dot(X: number[][], Y: number[][]) {
  const res: number[][] = []

  for (let i = 0; i < 8; i++) {
    const xRow = X[i % 8]
    res[i] = []
    for (let j = 0; j < 8; j++) {
      const yCol = Y.map(row => row[j % 8])
      const v = xRow.reduce((acc, cur, idx) => acc + cur * yCol[idx], 0)
      res[i][j] = v
    }
  }

  return res
}
