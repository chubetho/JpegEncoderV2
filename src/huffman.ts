import { acTables, dcTables, zigzag } from './constants'
import { useStream } from './stream'

export function generateCodes(t: HuffmanTable) {
  let code = 0
  for (let i = 0; i < 16; i++) {
    for (let j = t.offsets[i]; j < t.offsets[i + 1]; j++) {
      t.codes[j] = code
      code += 1
    }
    code <<= 1
  }
}

export function getBitLength(v: number) {
  let length = 0
  while (v > 0) {
    v >>= 1
    length += 1
  }
  return length
}

export function getCode(t: HuffmanTable, symbol: number) {
  let code = 0
  let codelength = 0

  for (let i = 0; i < 16; i++) {
    for (let j = t.offsets[i]; j < t.offsets[i + 1]; j++) {
      if (symbol !== t.symbols[j])
        continue

      code = t.codes[j]
      codelength = i + 1
      return { code, codelength }
    }
  }
}

export function getHuffmanData(img: Image) {
  const { data, writeBits } = useStream()
  const dcDiffs = [0, 0, 0]

  for (let i = 0; i < 3; i++) {
    if (!dcTables[i].set) {
      generateCodes(dcTables[i])
      dcTables[i].set = true
    }
    if (!acTables[i].set) {
      generateCodes(acTables[i])
      acTables[i].set = true
    }
  }

  for (let h = 0; h < img.metadata.blockHeight; h++) {
    for (let w = 0; w < img.metadata.blockWidth; w++) {
      const block = img.blocks[h * img.metadata.blockWidth + w]
      for (let i = 0; i < 3; i++) {
        const component = i === 0 ? block.Y : i === 0 ? block.Cb : block.Cr
        if (!encoderBlockComponent(component, dcDiffs, i, dcTables[i], acTables[i]))
          return []
      }
    }
  }

  return data

  function encoderBlockComponent(component: Int32Array, dcDiffs: number[], dcDiffIdx: number, dcTable: HuffmanTable, acTable: HuffmanTable) {
    // DC
    let coeff = Math.ceil(component[0]) - dcDiffs[dcDiffIdx]
    dcDiffs[dcDiffIdx] = component[0]

    let coeffLength = getBitLength(Math.abs(coeff))
    if (coeffLength > 11) {
      console.log('dc coefficient length > 11')
      return false
    }
    if (coeff < 0)
      coeff += (1 << coeffLength) - 1

    const dcCode = getCode(dcTable, coeffLength)
    if (!dcCode) {
      console.log('invalid dc')
      return false
    }
    writeBits(dcCode.code, dcCode.codelength)
    writeBits(coeff, coeffLength)

    // AC
    for (let i = 1; i < 64; i++) {
      let count = 0
      while (i < 64 && component[zigzag[i]] === 0) {
        count += 1
        i += 1
      }

      if (i === 64) {
        const acCode = getCode(acTable, 0x00)
        if (!acCode) {
          console.log('invalid ac')
          return false
        }
        writeBits(acCode.code, acCode.codelength)
        return true
      }

      while (count >= 16) {
        const acCode = getCode(acTable, 0xF0)
        if (!acCode) {
          console.log('invalid ac')
          return false
        }
        writeBits(acCode.code, acCode.codelength)
        count -= 16
      }

      coeff = component[zigzag[i]]
      coeffLength = getBitLength(Math.abs(coeff))
      if (coeffLength > 10) {
        console.log('ac coefficient length > 10')
        return false
      }
      if (coeff < 0)
        coeff += (1 << coeffLength) - 1

      const symbol = count << 4 | coeffLength
      const symbolCode = getCode(acTable, symbol)
      if (!symbolCode) {
        console.log('invalid ac')
        return false
      }

      writeBits(symbolCode.code, symbolCode.codelength)
      writeBits(coeff, coeffLength)
    }

    return true
  }
}
