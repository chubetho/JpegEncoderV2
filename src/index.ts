import * as fs from 'node:fs'
import { Buffer } from 'node:buffer'
import { useStream } from './stream'
import { acTableC, acTableY, dcTableC, dcTableY, qTableC, qTableY, zigzag } from './constants'
import { dct, dqt } from './transform'

const min = Math.min
const abs = Math.abs

function read(path: string) {
  const content = fs.readFileSync(path, 'ascii')
  const lines = content.split('\n')
  const [format, size, _maxColor, ...rest] = lines
  const [width, height] = size.trim().split(' ').map(Number)
  const maxColor = Number.parseInt(_maxColor)

  if (format !== 'P3')
    throw new Error('Invalid ppm')

  if (maxColor !== 255)
    throw new Error('Invalid ppm')

  const image = new Uint8Array(width * height * 3)
  let index = 0

  for (let l = 0; l < rest.length; l++) {
    const trim = rest[l].trim()
    if (trim[0] === '#')
      continue

    const row = trim.split(' ')
    for (let i = 0; i < row.length; i++) {
      if (!row[i])
        continue
      const v = Number.parseInt(row[i])
      if (Number.isNaN(v) || v < 0 || v > maxColor)
        throw new Error('Invalid ppm')

      image[index++] = v
    }
  }

  return {
    image,
    width,
    height,
  }
}

function rgb2y(r: number, g: number, b: number) { return +0.299 * r + 0.587 * g + 0.114 * b }
function rgb2cb(r: number, g: number, b: number) { return -0.16874 * r - 0.33126 * g + 0.5 * b }
function rgb2cr(r: number, g: number, b: number) { return +0.5 * r - 0.41869 * g - 0.08131 * b }

function generateCodes(t: HuffmanTable) {
  let code = 0
  for (let i = 0; i < 16; i++) {
    for (let j = t.offsets[i]; j < t.offsets[i + 1]; j++) {
      t.codes[j] = code
      code += 1
    }
    code <<= 1
  }
  t.set = true
}

function getBitLength(v: number) {
  let length = 0
  while (v > 0) {
    v >>= 1
    length += 1
  }
  return length
}

function getCode(t: HuffmanTable, symbol: number) {
  if (!t.set)
    throw new Error('table not set')

  let code = 0
  let codelength = 0

  for (let i = 0; i < 16; i++) {
    for (let j = t.offsets[i]; j < t.offsets[i + 1]; j++) {
      if (symbol !== t.symbols[j])
        continue

      code = t.codes[j]
      codelength = i + 1
      return { code, codelength }
    }
  }
}

interface Config {
  subsampling?: boolean
}

export function encoder(path: string, config?: Config) {
  const { subsampling }: Config = {
    subsampling: true,
    ...config,
  }

  !dcTableY.set && generateCodes(dcTableY)
  !dcTableC.set && generateCodes(dcTableC)
  !acTableY.set && generateCodes(acTableY)
  !acTableC.set && generateCodes(acTableC)

  const { image: pixels, width, height } = read(path)
  const { getData, writeByte, writeWord, writeBits } = useStream()

  // SOI
  writeWord(0xFFD8)

  // APP0
  writeWord(0xFFE0)
  writeWord(16)
  writeByte(0x4A) // J
  writeByte(0x46) // I
  writeByte(0x49) // F
  writeByte(0x46) // I
  writeByte(0)
  writeByte(1)
  writeByte(1)
  writeByte(0)
  writeWord(1)
  writeWord(1)
  writeByte(0)
  writeByte(0)

  // DQT
  writeWord(0xFFDB)
  writeWord(2 + 2 * 65)
  writeByte(0)
  for (let i = 0; i < 64; i++)
    writeByte(qTableY[zigzag[i]])
  writeByte(1)
  for (let i = 0; i < 64; i++)
    writeByte(qTableC[zigzag[i]])

  // SOF
  writeWord(0xFFC0)
  writeWord(17)
  writeByte(8) // precision
  writeWord(height)
  writeWord(width)

  writeByte(3) // components

  writeByte(1) // Y
  writeByte(subsampling ? 0x22 : 0x11)
  writeByte(0) // QtY

  writeByte(2) // Cb
  writeByte(0x11)
  writeByte(1) // QtC

  writeByte(3) // Cr
  writeByte(0x11)
  writeByte(1) // QtC

  // DHT
  const writeHuffmanTable = (type: number, id: number, table: HuffmanTable) => {
    writeWord(0xFFC4)
    writeWord(19 + table.offsets[16])
    writeByte(type << 4 | id)

    for (let i = 0; i < 16; i++)
      writeByte(table.offsets[i + 1] - table.offsets[i])

    for (let i = 0; i < 16; i++) {
      for (let j = table.offsets[i]; j < table.offsets[i + 1]; j++)
        writeByte(table.symbols[j])
    }
  }

  writeHuffmanTable(0, 0, dcTableY)
  writeHuffmanTable(0, 1, dcTableC)
  writeHuffmanTable(1, 0, acTableY)
  writeHuffmanTable(1, 1, acTableC)

  // SOS
  writeWord(0xFFDA)
  writeWord(12)
  writeByte(3) // components

  writeByte(1) // Y
  writeByte(0x00) // Huffmantable

  writeByte(2) // Cb
  writeByte(0x11) // Huffmantable

  writeByte(3) // Cr
  writeByte(0x11) // Huffmantable

  writeByte(0)
  writeByte(63)
  writeByte(0)

  // Data
  let lastYDc = 0
  let lastCbDc = 0
  let lastCrDc = 0

  const encode = (comp: Int16Array, qTable: Uint8Array, dcDiff: number, dcTable: HuffmanTable, acTable: HuffmanTable) => {
    dct(comp)
    dqt(comp, qTable)

    // DC
    let coeff = comp[0] - dcDiff

    let coeffLength = getBitLength(abs(coeff))
    if (coeffLength > 11)
      throw new Error('dc coefficient length > 11')

    if (coeff < 0)
      coeff += (1 << coeffLength) - 1

    const coeffLengthCode = getCode(dcTable, coeffLength)
    if (!coeffLengthCode)
      throw new Error('invalid dc')

    writeBits(coeffLengthCode.code, coeffLengthCode.codelength)
    writeBits(coeff, coeffLength)

    // AC
    for (let i = 1; i < 64; i++) {
      let count = 0
      while (i < 64 && comp[zigzag[i]] === 0) {
        count++
        i++
      }

      if (i === 64) {
        const eob = getCode(acTable, 0x00)
        if (!eob)
          throw new Error('invalid ac')

        writeBits(eob.code, eob.codelength)
        continue
      }

      while (count >= 16) {
        const xf0 = getCode(acTable, 0xF0)
        if (!xf0)
          throw new Error('invalid ac')

        writeBits(xf0.code, xf0.codelength)
        count -= 16
      }

      coeff = comp[zigzag[i]]
      coeffLength = getBitLength(abs(coeff))
      if (coeffLength > 10)
        throw new Error('ac coefficient length > 10')

      if (coeff < 0)
        coeff += (1 << coeffLength) - 1

      const symbol = count << 4 | coeffLength
      const symbolCode = getCode(acTable, symbol)
      if (!symbolCode)
        throw new Error('invalid ac')

      writeBits(symbolCode.code, symbolCode.codelength)
      writeBits(coeff, coeffLength)
    }

    return comp[0]
  }

  const maxWidth = width - 1
  const maxHeight = height - 1
  const mcuSize = subsampling ? 16 : 8

  const Y = new Int16Array(64)
  const Cb = new Int16Array(64)
  const Cr = new Int16Array(64)

  let YCount = 0

  for (let mcuY = 0; mcuY < height; mcuY += mcuSize) {
    for (let mcuX = 0; mcuX < width; mcuX += mcuSize) {
      for (let blockY = 0; blockY < mcuSize; blockY += 8) {
        for (let blockX = 0; blockX < mcuSize; blockX += 8) {
          for (let deltaY = 0; deltaY < 8; deltaY++) {
            let column = min(mcuX + blockX, maxWidth)
            const row = min(mcuY + blockY + deltaY, maxHeight)
            for (let deltaX = 0; deltaX < 8; deltaX++) {
              const pixelPos = row * width + column
              if (column < maxWidth)
                column++
              const r = pixels[3 * pixelPos]
              const g = pixels[3 * pixelPos + 1]
              const b = pixels[3 * pixelPos + 2]

              Y[deltaY * 8 + deltaX] = rgb2y(r, g, b) - 128
              if (!subsampling) {
                Cb[deltaY * 8 + deltaX] = rgb2cb(r, g, b)
                Cr[deltaY * 8 + deltaX] = rgb2cr(r, g, b)
              }
            }
          }

          lastYDc = encode(Y, qTableY, lastYDc, dcTableY, acTableY)
          YCount++

          if (subsampling) {
            for (let deltaY = 0; deltaY < 8; deltaY++) {
              const row = min(mcuY + 2 * deltaY, maxHeight)
              let column = mcuX
              let pixelPos = (row * width + column) * 3
              const rowStep = row < maxHeight ? 3 * width : 0
              let columnStep = column < maxWidth ? 3 : 0
              for (let deltaX = 0; deltaX < 8; deltaX++) {
                const right = pixelPos + columnStep
                const down = pixelPos + rowStep
                const downRight = pixelPos + columnStep + rowStep

                const r = pixels[pixelPos] + pixels[right] + pixels[down] + pixels[downRight]
                const g = pixels[pixelPos + 1] + pixels[right + 1] + pixels[down + 1] + pixels[downRight + 1]
                const b = pixels[pixelPos + 2] + pixels[right + 2] + pixels[down + 2] + pixels[downRight + 2]
                Cb[deltaY * 8 + deltaX] = rgb2cb(r, g, b) / 4
                Cr[deltaY * 8 + deltaX] = rgb2cr(r, g, b) / 4

                pixelPos += 2 * 3
                column += 2

                if (column >= maxWidth) {
                  columnStep = 0
                  pixelPos = ((row + 1) * width - 1) * 3
                }
              }
            }
            if (YCount === 4) {
              lastCbDc = encode(Cb, qTableC, lastCbDc, dcTableC, acTableC)
              lastCrDc = encode(Cr, qTableC, lastCrDc, dcTableC, acTableC)
              YCount = 0
            }
          }
          else {
            lastCbDc = encode(Cb, qTableC, lastCbDc, dcTableC, acTableC)
            lastCrDc = encode(Cr, qTableC, lastCrDc, dcTableC, acTableC)
          }
        }
      }
    }
  }

  // EOI
  writeWord(0xFFD9)
  return Buffer.from(getData())
}
