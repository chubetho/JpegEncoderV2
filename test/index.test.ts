import { writeFileSync } from 'node:fs'
import { expect, it } from 'bun:test'
import { encoder } from '../src'

it('small_sub', () => {
  const buffer = encoder('src/assets/small.ppm', { subsampling: true })
  expect(buffer).toMatchSnapshot()
  writeFileSync('src/output/small_sub.jpg', buffer)
})

it('small', () => {
  const buffer = encoder('src/assets/small.ppm', { subsampling: false })
  expect(buffer).toMatchSnapshot()
  writeFileSync('src/output/small.jpg', buffer)
})

it('medium_sub', () => {
  const buffer = encoder('src/assets/medium.ppm', { subsampling: true })
  expect(buffer).toMatchSnapshot()
  writeFileSync('src/output/medium_sub.jpg', buffer)
})

it('medium', () => {
  const buffer = encoder('src/assets/medium.ppm', { subsampling: false })
  expect(buffer).toMatchSnapshot()
  writeFileSync('src/output/medium.jpg', buffer)
})

it.skip('big_sub', () => {
  const buffer = encoder('src/assets/big.ppm', { subsampling: true })
  writeFileSync('src/output/big_sub.jpg', buffer)
})

it.skip('big', () => {
  const buffer = encoder('src/assets/big.ppm', { subsampling: false })
  writeFileSync('src/output/big.jpg', buffer)
})

it.skip('nasa_sub', () => {
  const buffer = encoder('src/assets/nasa.ppm', { subsampling: true })
  writeFileSync('src/output/nasa_sub.jpg', buffer)
})

it.skip('nasa', () => {
  const buffer = encoder('src/assets/nasa.ppm', { subsampling: false })
  writeFileSync('src/output/nasa.jpg', buffer)
})
