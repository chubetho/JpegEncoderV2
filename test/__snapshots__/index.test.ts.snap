// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`foo 1`] = `
[
  225,
  255,
  0,
  255,
  0,
  255,
  0,
  255,
  0,
  255,
  0,
  255,
  0,
  249,
  15,
  255,
  0,
  255,
  0,
  255,
  0,
  255,
  0,
  255,
  0,
  255,
  0,
  63,
  255,
  0,
  255,
  0,
  255,
  0,
  255,
  0,
  207,
  255,
  0,
  255,
  0,
  255,
  0,
  255,
  0,
  255,
  0,
  241,
  255,
  0,
  255,
  0,
  255,
  0,
  255,
  0,
  255,
  0,
  15,
  255,
  0,
  255,
  0,
  255,
  0,
  255,
  0,
  243,
  255,
  0,
  255,
  0,
  255,
  0,
  255,
  0,
  255,
  0,
  253,
  63,
  255,
  0,
  255,
  0,
  255,
  0,
  255,
  0,
  195,
  255,
  0,
  255,
  0,
  255,
  0,
  255,
  0,
  253,
  255,
  0,
  255,
  0,
  255,
  0,
  255,
  0,
  255,
  0,
  255,
  0,
  15,
  255,
  0,
  255,
  0,
  255,
  0,
  255,
  0,
  241,
  255,
  0,
  255,
  0,
  255,
  0,
  255,
  0,
  255,
  0,
]
`;
